compiler/z/codegen/ControlFlowEvaluator.cpp

```diff
@@ -2724,99 +2724,50 @@ TR::Register *
    TR::Node *falseValueNode = node->getThirdChild();
-
-   TR::Register *trueValueReg = cg->evaluate(trueValueNode);
-   TR::Register *falseValueReg = cg->evaluate(falseValueNode);
-
-   TR::Register *vectorSelReg = cg->allocateRegister(TR_VRF);
-   TR::Register *returnReg = cg->allocateRegister(TR_FPR);
-
-   if (conditionNode->getOpCode().isCompareDouble())
-      {
-      TR::Node *firstCmpValue = conditionNode->getFirstChild();
-      TR::Node *secondCmpValue = conditionNode->getSecondChild();
-
-      TR::Register *firstCmpReg = cg->evaluate(firstCmpValue);
-      TR::Register *secondCmpReg = cg->evaluate(secondCmpValue);
-
-      TR::InstOpCode::Mnemonic compareOp = TR::InstOpCode::BAD;
-      TR::ILOpCodes ifOp = conditionNode->getOpCodeValue();
-
-      switch (ifOp)
-         {
-         case TR::dcmpne:
-         case TR::dcmpneu:
-         case TR::dcmpeq:
-         case TR::dcmpequ:
-            compareOp = TR::InstOpCode::VFCE;
-            break;
-         case TR::dcmplt:
-         case TR::dcmpgt:
-         case TR::dcmpleu:
-         case TR::dcmpgeu:
-            compareOp = TR::InstOpCode::VFCH;
-            break;
-         case TR::dcmpltu:
-         case TR::dcmpgtu:
-         case TR::dcmple:
-         case TR::dcmpge:
-            compareOp = TR::InstOpCode::VFCHE;
-            break;
-         }
-
-      // To handle unordered compares, we switch the compare operation, since the vector compare instructions return false for unordered.
-      // For example, if the original compare is dcmpgeu:
-      //    (x >= y [true if unordered]) ? a : b
-      //    switch to (x < y) ? b : a;
-      // So that only if x < y, do we select the "false value"
-
-      if (ifOp == TR::dcmpgeu || ifOp == TR::dcmpgtu || ifOp == TR::dcmplt || ifOp == TR::dcmple)
-         generateVRRcInstruction(cg, compareOp, node, vectorSelReg, secondCmpReg, firstCmpReg, 1, 0, 3);
-      else
-         generateVRRcInstruction(cg, compareOp, node, vectorSelReg, firstCmpReg, secondCmpReg, 1, 0, 3);
-
-      // The above operation returns 0's for not equals, so we must swap the true and false values to select the correct "true value" for cmpne operations
-      // Swap the values for unordered comparisons also (see above)
-      if (ifOp == TR::dcmpne || conditionNode->getOpCode().isCompareTrueIfUnordered())
-         generateVRReInstruction(cg, TR::InstOpCode::VSEL, node, returnReg, falseValueReg, trueValueReg, vectorSelReg);
-      else
-         generateVRReInstruction(cg, TR::InstOpCode::VSEL, node, returnReg, trueValueReg, falseValueReg, vectorSelReg);
-
-      cg->stopUsingRegister(firstCmpReg);
-      cg->stopUsingRegister(secondCmpReg);
-      }
-   else
+   TR::Register *resultReg = cg->gprClobberEvaluate(trueValueNode);
+   TR::Register *conditionReg = cg->evaluate(conditionNode);
+   TR::Register *falseValReg = cg->evaluate(falseValueNode);
+   if (cg->comp()->target().cpu.getSupportsArch(TR::CPU::z13) && node->getOpCode().isDouble())
```

Comment:
Can this also work for floats using `2` for M4 field in `VFCE`?

Reply:
My initial changes was trying to use VFCE in short format for float, but I was hitting illegal instruction for `VSEL` which works completely fine with the double format. VSEL does not have short format. I have not done detailed investigation using GDB, why it was happening, but my instinct is that the result register in VFCE is marked as short format, and when I reuse that in the `VSEL` it fails with illegal instruction.

Reply:
Looking at the documentation for VSEL in PoPs I don't see why it would cause SIGILL. It just selects bits from one register and places then into another. It doesn't even have a notion of short or long format. In either way this could be a small enhancement in the future for some beginner to take a look. We need to fix the immediate problem at hand so I'm going to go ahead and merge this and open up and issue to further investigate support for floats. Perhaps someone new to the codegen can tackle that.

Reply:
Opened up #5002 to further investigate.